pkg/plugins/runtime/k8s/controllers/gateway_controller.go

```diff
@@ -0,0 +1,200 @@
+//go:build gateway
+// +build gateway
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	kube_apps "k8s.io/api/apps/v1"
+	kube_core "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kube_runtime "k8s.io/apimachinery/pkg/runtime"
+	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_ctrl "sigs.k8s.io/controller-runtime"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	kube_controllerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	core_ca "github.com/kumahq/kuma/pkg/core/ca"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	k8s_common "github.com/kumahq/kuma/pkg/plugins/common/k8s"
+	mesh_k8s "github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/webhooks/injector"
+	util_k8s "github.com/kumahq/kuma/pkg/util/k8s"
+)
+
+// GatewayController reconciles a gateway.kuma.io object
+type GatewayController struct {
+	kube_client.Client
+	Reader kube_client.Reader
+	Log    logr.Logger
+
+	Scheme          *kube_runtime.Scheme
+	Converter       k8s_common.Converter
+	CaManagers      core_ca.Managers
+	SystemNamespace string
+	Injector        injector.KumaInjector
+	ResourceManager manager.ResourceManager
+}
+
+func (r *GatewayController) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	k8sGateway := &mesh_k8s.Gateway{}
+	if err := r.Get(ctx, req.NamespacedName, k8sGateway); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			mesh := MeshFor(k8sGateway.Annotations)
+			coreName := util_k8s.K8sNamespacedNameToCoreName(req.Name, req.Namespace)
+
+			err := r.ResourceManager.Delete(ctx, core_mesh.NewGatewayResource(), store.DeleteByKey(coreName, mesh))
```

Comment:
This was in the `mesh_controller`, which is the only controller we have that watches our own resources.

It's not clear to me whether we need it. If it's necessary for `Mesh`es it might very well be necessary for `Gateway`s since this is kind of deletion by bypassing the kuma store, which could theoretically create problems for any resource.

Owner references take care of the deletion of any dependent k8s resources created by this controller.